dataplane/api-server/src/netutils.rs

```diff
@@ -5,59 +5,89 @@ SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
-
-    Ok(device)
+// pub fn if_nametoindex(ifname: String) -> Result<u32, Error> {
+//     let ifname_c = CString::new(ifname)?;
+//     let ifindex = unsafe { libc_if_nametoindex(ifname_c.into_raw()) };
+//     Ok(ifindex)
+// }
+
+/// Returns an net interface index for a Ipv4 address, like `ip route get to`
+pub fn if_index_for_routing_ip(ip_addr: Ipv4Addr) -> Result<u32, Error> {
+    // run the linux command "ip route" to get the device's index responsible for
+    // routing the given Ipv4 address.
+    let mut socket = Socket::new(NETLINK_ROUTE).unwrap();
+    let _port_number = socket.bind_auto().unwrap().port_number();
+    socket.connect(&SocketAddr::new(0, 0)).unwrap();
+
+    let mut nl_hdr = NetlinkHeader::default();
+    nl_hdr.flags = NLM_F_REQUEST | NLM_F_DUMP_FILTERED;
+
+    // construct RouteMessage
+    let route_header = RouteHeader {
+        address_family: AddressFamily::Inet,
+        flags: RouteFlags::LookupTable,
+        destination_prefix_length: 32,
+        table: RouteHeader::RT_TABLE_MAIN,
+        ..Default::default()
+    };
+    let route_attribute = RouteAttribute::Destination(RouteAddress::Inet(ip_addr));
+    let mut route_message = RouteMessage::default();
+    route_message.attributes = vec![route_attribute];
+    route_message.header = route_header;
+
+    let no_device_err: String = format!("no device found to route {}", ip_addr);
+    let con_packet_err: String = "construct packet failed".to_string();
+    let nl_send_msg_err: String = "netlink send message failed".to_string();
+    let nl_recv_msg_err: String = "netlink receive message failed".to_string();
+
+    let mut packet = NetlinkMessage::new(
+        nl_hdr,
+        NetlinkPayload::from(RouteNetlinkMessage::GetRoute(route_message)),
+    );
+    packet.finalize();
+    let mut buf = vec![0; packet.header.length as usize];
+    // check packet
+    if buf.len() != packet.buffer_len() {
+        return Err(Error::msg(con_packet_err));
+    }
+    packet.serialize(&mut buf[..]);
+    socket
+        .send(&buf[..], 0)
+        .map_err(|_| Error::msg(nl_send_msg_err))?;
+
+    let mut receive_buffer = vec![0; 4096];
+    socket
+        .recv(&mut &mut receive_buffer[..], 0)
+        .map_err(|_| Error::msg(nl_recv_msg_err))?;
+
+    let bytes = &receive_buffer[..];
+
+    // extract returned RouteNetLinkMessage
+    let (_, payload) = <NetlinkMessage<RouteNetlinkMessage>>::deserialize(bytes)
+        .map_err(|_| Error::msg(no_device_err.clone()))?
+        .into_parts();
+    match payload {
+        NetlinkPayload::InnerMessage(RouteNetlinkMessage::NewRoute(v)) => {
+            if let Some(RouteAttribute::Oif(idex_if)) = v
+                .attributes
+                .iter()
+                .find(|attr| matches!(attr, RouteAttribute::Oif(_)))
+            {
+                return Ok(*idex_if);
+            }
+            Err(Error::msg(no_device_err.clone()))
+        }
+
+        _ => Err(Error::msg(no_device_err)),
+    }
 }
```

Comment:
It would ultimately be good to add tests here, albeit it's pretty obvious given the use of netlink here that tests would be relegated to something more in the integration testing kind of space. We'll need to continue thinking about how we should do testing for components like this :thinking:

Reply:
so now I just add unit test for this func?

Reply:
> It would ultimately be good to add tests here, albeit it's pretty obvious given the use of netlink here that tests would be relegated to something more in the integration testing kind of space. We'll need to continue thinking about how we should do testing for components like this 🤔

so now I just add unit test for this func?

Reply:
It's difficult to make effective tests here, since we're basically just exercising netlink :thinking: Actually so that we don't have to add a long tail to this PR, and given the lack of risk to the project because the project is effectively an experimental sandbox, I think I'll back up and take a holistic look at testing for a later iteration. We can move forward as-is here and I'll follow up.